libcnb/src/runtime.rs

```diff
@@ -140,21 +140,38 @@ pub fn libcnb_runtime_detect<B: Buildpack>(
     #[cfg(not(feature = "trace"))]
     let trace_error = |_: &dyn std::error::Error| {};
 
-    let stack_id: StackId = env::var("CNB_STACK_ID")
-        .map_err(Error::CannotDetermineStackId)
-        .and_then(|stack_id_string| stack_id_string.parse().map_err(Error::StackIdError))
-        .inspect_err(|err| trace_error(err))?;
-
     let platform = B::Platform::from_path(&args.platform_dir_path)
         .map_err(Error::CannotCreatePlatformFromPath)
         .inspect_err(|err| trace_error(err))?;
 
     let build_plan_path = args.build_plan_path;
 
+    let target_os = env::var("CNB_TARGET_OS")
```

Comment:
I'm not sure I quite follow the spec about these env vars:
https://github.com/buildpacks/spec/blob/main/buildpack.md#targets

It says:

> The following environment variables MUST be set by the lifecycle during the detect and build phases to describe the target runtime image, if inputs are provided.

However:
1. What does "if inputs are provided" mean? Does it mean "some of these might not be set if they are not set in the builder image"? (In which case, do all of these need to be `Option<String>`? Also, are any of these automatically derived from the builder image, even if they are not set as labels there? eg there is mention of [this bionic backwards compat thing](https://github.com/buildpacks/spec/blob/main/buildpack.md#buildpacktoml-toml-stacks-array), but it presumably only matches the "bionic" case exactly, rather than any other builder IDs?
2. If the spec says MUST why does it then say "optional" for some of the env vars, like `CNB_TARGET_ARCH_VARIANT`

I guess I'm not clear how the tests are passing, given that our build images don't set the new labels yet, and only the legacy `io.buildpacks.stack.id` label and `CNB_STACK_ID` env var?
https://github.com/heroku/base-images/blob/0874d827bb161c4df57580840010063122e01130/heroku-22-cnb-build/Dockerfile#L12-L13

I also tried reading https://github.com/buildpacks/spec/blob/main/platform.md#target-data but that didn't really help either.

Reply:
> Are any of these automatically derived from the builder image, even if they are not set as labels there? eg there is mention of [this bionic backwards compat thing](https://github.com/buildpacks/spec/blob/main/buildpack.md#buildpacktoml-toml-stacks-array), but it presumably only matches the "bionic" case exactly, rather than any other builder IDs?

I also found this backwards compat matrix (which was linked from one of the RFC or spec issues, I forget where) that suggests maybe the env vars are derived from the "image config" after all?
https://user-images.githubusercontent.com/6894483/222824654-09bd9c3f-26a8-44a1-90c6-f78d37bc4213.png

But does "image config" mean the manually provided labels in the `Dockerfile`, or the OCI metadata that automatically gets set when `docker build`ing the build image?

Reply:
My understanding is that the `CNB_TARGET_*` environment variables will be filled from various sources if not explicitly set. Examples:

- [From systemd](https://github.com/buildpacks/lifecycle/blob/751c7a2737be2e6a8f66bfcc2e0cf0058661d6ff/platform/target_data.go#L60-L73)
- [From the run image metadata](https://github.com/buildpacks/lifecycle/blob/751c7a2737be2e6a8f66bfcc2e0cf0058661d6ff/platform/target_data.go#L27-L56)

Reply:
Agreed, the spec is very unclear there. Lifecycle also sets `CNB_TARGET_DISTRO_NAME` to an empty string instead of not setting it at all when no name can be found. Unclear if that is intentional or not. Since empty strings in Go are are treated as Nil, this might a Go (yet again) leaking into the spec.

Reply:
Based on the fact that this works in practice with `CNB_TARGET_OS` and `CNB_TARGET_ARCH` being required and lifecycle actively setting these values even if not provided, I think we can continue with the PR as-is and have to raise these questions to upstream separately.

Reply:
> I think we can continue with the PR as-is and have to raise these questions to upstream separately.

Yeah agree it needn't block here - after this merges it would be good to make sure we track (via a TODO or GUS WI or GH issue in this repo) raising these upstream sometime soon before it fades too much from our memories :-)

Reply:
Relevant spec Slack thread with upstream: https://cloud-native.slack.com/archives/C033DV9EBDF/p1707493563612439